### ISSUE TITLE
Drawing order when multiple units occupy the same hex

### ISSUE THREAD

**Issue by BLOODWOLF333** [MegaMek/megamek#180]
I feel very strongly that when multiple units occupy the same hex that either the tallest/highest units should draw over all those below them or at least the opponents units would draw above the player's own.  Infantry > vehicles > mechs > aerospace fighters.  This is a bird's eye view point game so, I would think that makes sense that the highest unit would draw over all those below it.  Currently it's almost like MM can hide targets from you if you did not realize or see that the other player had moved one of their units into the same hex as say two of your tanks.  Yes the name tags are visible but, can be covered up by the player's unit's name tags if several player units are grouped together.  I am human I do make mistakes and sometimes do not pay attention but, it would help me at least to be able to see all of my opponents units in LOS during the firing phase and not have any of them hidden under my own.
![ingame](https://cloud.githubusercontent.com/assets/16950832/15061376/0d226440-12fa-11e6-954f-55fa3967674d.png)
 There is an enemy Panther is hex 2527 under the two condor tanks with it's name tag partially covered up by Condor #32.  I usually play zoomed out to the level before the hex coords disappear and this just works against me.

**Comment by SJuliez** [comment]
That feels like two separate issues. The ordering of the units can be done (nothing wrong with it for me).

The other point is that it's fairly easy to miss units when theres a big brawl going on. In the movement phase they can move around and one can lose track and in the firing phase one might miss a good target. Maybe we could add a function to highlight all hexes with enemy units in them as well as screen border arrows towards off-screen hexes with enemy units.

Someone already suggested adding a clear player color mark to all unit labels (or, maybe give enemy ones a read border?). That might also help.

**Comment by Akjosch** [comment]
Another suggestion: Have a toggable display of all / own / enemy / none units. Do we use TAB for anything yet? It sounds like a natural key to switch between the display modes.

**Comment by SJuliez** [comment]
For me, it always felt like TAB should be the real key for "next unit" :)

**Comment by arlith** [comment]
I'm reluctant to change the draw order of units.  I don't think a simple ordering is sufficient.  You've got superheavy vehicles and ultralight mechs, infantry with field pieces, etc.  Plus, the icons themselves aren't restricted based on weight.  You could have an Omega with a Locust icon.

I think Juliez' comment gets to the core if the issue: how we indicate multiple units in a hex.  I personally don't think the drawing order is all that important.

**Comment by BLOODWOLF333** [comment]
Then what about ordering them based off height?  Top down view game a mech should stick up past a tank or could be drawn over any tanks and would be more visible is how I'm rationalizing it.  The tanks are wide and would still be visible so that both sides could identify what was going on in the hex from my screenshot instead of the panther almost completely covered up.

**Comment by BLOODWOLF333** [comment]
Do units have base types?  Like are protos and super heavies just a base type of mek which this could be tied to?

**Comment by Akjosch** [comment]
The hierarchy is a bit complicated and sadly doesn't always make the most sense.

* Entity
 * Aero
   * ConvFighter
     * FixedWingSupport
     * FighterSquadron
     * Jumpship
       * SpaceStation
       * Warship
    * SmallCraft
      * Dropship
    * TeleMissile
  * Infantry
    * BattleArmor
    * EjectedCrew
      * MechWarrior
  * Mech
    * BipedMech
      * ArmlessMech
      * LandAirMech
    * QuadMech
    * TripodMech
  * ProtoMech
  * Tank
    * GunEmplacement
    * SuperHeavyTank
    * SupportTank
      * LargeSupportTank
    * VTOL
      * SupportVTOL

**Comment by BLOODWOLF333** [comment]
Ah ok.  Infantry>tank>protomech>mech>aero/VTOL.  But, I guess then the opponent could move their battlearmor into the same hex as a vehicle and you wouldnt see them and people could bring up the same issue I had with mechs covered by vehicles.  I see drawing order may not be the fix all.  But, could help?

**Comment by BLOODWOLF333** [comment]
So I realize I screwed myself over on that game by holding the screen over my opponents ASFs instead of letting the screen swoop over to my units as they were firing and I would have seen the cyan TH text over the hex telling me an enemy unit was there.  But that feature has been made optional I hear in the next release so I have no qualms about it.  I guess I am the kinda player that needs the instant teleport to "force" me to look at my units before I fire them, others probably pay more attention.  Was suggested on IRC to have a red or whatever color the team is, box appear above or inside the top edge of the hex when another unit shares the same hex.  I think now maybe it would just be better to have opponent unit's name tags background be red or their team color which will make it even more obvious where the opponents units are and not just during the firing phase with the TH text.

**Comment by BLOODWOLF333** [comment]
Example
![ingame1](https://cloud.githubusercontent.com/assets/16950832/15095761/11b699a8-149f-11e6-8c97-a2e0ceca4dc5.png)

**Comment by arlith** [comment]
I finally gave in and implemented a draw order.  Previously, the draw order was, "whatever unit was put there first."  This seemed like it could lead to odd situations where in one hex, infantry are the bottom, and in another they are on top.  Defining a draw order adds some consistency.  What I did was change the `EntitySprite` and `IsometricEntitySprite` collections to `PriorityQueue`s.  I then had `Sprite` extended `Comparable` and I added a `Sprite.getSpritePriority`.  `Sprite`'s compareTo implementation uses the priority, which can then be overridden in subclasses.  For `EntitySprite` and `IsometricEntitySprite`, I used a new `Entity.getSpriteDrawPriority`, which I then defined for `Aeror,` `VTOL`, `Mech`, `Tanks, `BattleArmor`, and `Infantry` in that order.

Additionally, I made it so (optionally) a colored border is drawn around unit labels, based on the player color of the owner.
